scipy/sparse/csgraph/_flow.pyx

```diff
@@ -680,3 +709,881 @@ cdef ITYPE_t[:] _dinic(
+    the minimum cost flow problem consists of finding a flow whose cost (which is
+    defined as the scalar product of the edges' flows and costs) is minimal.
+
+    To solve the problem, we use Network Simplex [2]_.
+    The time complexity of the former :math:`O(|V|^2\,|E|\,log(|V|\,C))`.
+
+    The minimum cost flow problem is usually defined with real valued capacities,
+    edge costs and vertex demands but we require that all capacities are integral.
+    When dealing with rational capacities, or capacities belonging to
+    :math:`x\mathbb{Q}` for some fixed :math:`x \in \mathbb{R}`, it is possible
+    to reduce the problem to the integral case by scaling all capacities
+    accordingly.
+
+    Solving a minimum cost flow problem has many practical uses such as
+    in assignment problem, covers and matching in bipartite graphs [3]_.
+
+    We have ported the Python implementation of NetworkX [4]_ to Cython.
+
+    References
+    ----------
+
+    .. [1] https://en.wikipedia.org/wiki/Minimum-cost_flow_problem#Definition
+    .. [2] Orlin, J.B. A polynomial time primal network simplex algorithm
+           for minimum cost flows. Mathematical Programming 78,
+           109-129 (1997).
+    .. [3] https://en.wikipedia.org/wiki/Network_simplex_algorithm#Applications
+    .. [4] A Hagberg, D Schult, P Swart, Exploring Network Structure, Dynamics,
+           and Function using NetworkX in Proceedings of the 7th Python in
+           Science conference (SciPy 2008), G Varoquaux, T Vaught, J Millman
+           (Eds.), pp. 11-15
+
+    Examples
+    --------
+    >>> from scipy.sparse import csr_matrix
+    >>> from scipy.sparse.csgraph import minimum_cost_flow
+    >>> A = csr_matrix([[0, 4, 10, 0],
+    ...                 [0, 0, 0, 9],
+    ...                 [0, 0, 0, 5],
+    ...                 [0, 0, 0, 0]])
+    >>> demand = [-5, 0, 0, 5]
+    >>> cost = [3, 6, 1, 2]
+    >>> result = minimum_cost_flow(A, demand, cost)
+    >>> result.flow_cost
+    24
+    >>> result.flow.toarray()
+    array([[0, 4, 1, 0],
+        [-4, 0, 0, 4],
+        [-1, 0, 0, 1],
+        [0, -4, -1, 0]], dtype=int32)
+    """
+    demand = np.asarray(demand)
+    cost = np.asarray(cost)
+    if not isspmatrix_csr(csgraph):
+        raise TypeError("graph must be in CSR format")
+    if not issubclass(csgraph.dtype.type, np.integer):
+        raise ValueError("graph capacities must be integers")
+    elif csgraph.dtype != ITYPE:
+        csgraph = csgraph.astype(ITYPE)
+    if not issubclass(demand.dtype.type, np.integer):
+        raise ValueError("vertex demands must be integers")
+    elif demand.dtype != ITYPE:
+        demand = demand.astype(ITYPE)
+    if not issubclass(cost.dtype.type, np.integer):
+        raise ValueError("edge costs must be integers")
+    elif cost.dtype != ITYPE:
+        cost = cost.astype(ITYPE)
+
+    n_verts = csgraph.indptr.shape[0] - 1
+    n_edges = cost.shape[0]
```

Comment:
Another cheap check could be to ensure that `demand` has length `n_verts`. And that `n_edges` matches up with the length of `csgraph.data` (or `csgraph.nnz`).